File: c-algorithms/src/matrix/matrix-multiplication/matrix-multiplication.c
#include <stdlib.h>
#include <emscripten.h>
#include "shared/shared.h"
#include "matrix/matrix-multiplication/matrix-multiplication.h";

EMSCRIPTEN_KEEPALIVE
Matrix_u64* matrixMultiplication_u64_c(Matrix_u64* matrixA, Matrix_u64* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_u64* c = makeMatrix_u64(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            u64 sum = 0;
            for (int k = 0 ; k < m; k++) {
                u64 Aik = matrixA->data[i][k];
                u64 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_i64* matrixMultiplication_i64_c(Matrix_i64* matrixA, Matrix_i64* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_i64* c = makeMatrix_i64(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            i64 sum = 0;
            for (int k = 0 ; k < m; k++) {
                i64 Aik = matrixA->data[i][k];
                i64 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_f64* matrixMultiplication_f64_c(Matrix_f64* matrixA, Matrix_f64* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_f64* c = makeMatrix_f64(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            f64 sum = 0;
            for (int k = 0 ; k < m; k++) {
                f64 Aik = matrixA->data[i][k];
                f64 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_f32* matrixMultiplication_f32_c(Matrix_f32* matrixA, Matrix_f32* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_f32* c = makeMatrix_f32(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            f32 sum = 0;
            for (int k = 0 ; k < m; k++) {
                f32 Aik = matrixA->data[i][k];
                f32 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_u32* matrixMultiplication_u32_c(Matrix_u32* matrixA, Matrix_u32* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_u32* c = makeMatrix_u32(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            u32 sum = 0;
            for (int k = 0 ; k < m; k++) {
                u32 Aik = matrixA->data[i][k];
                u32 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_i32* matrixMultiplication_i32_c(Matrix_i32* matrixA, Matrix_i32* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_i32* c = makeMatrix_i32(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            i32 sum = 0;
            for (int k = 0 ; k < m; k++) {
                i32 Aik = matrixA->data[i][k];
                i32 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_u16* matrixMultiplication_u16_c(Matrix_u16* matrixA, Matrix_u16* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_u16* c = makeMatrix_u16(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            u16 sum = 0;
            for (int k = 0 ; k < m; k++) {
                u16 Aik = matrixA->data[i][k];
                u16 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_i16* matrixMultiplication_i16_c(Matrix_i16* matrixA, Matrix_i16* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_i16* c = makeMatrix_i16(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            i16 sum = 0;
            for (int k = 0 ; k < m; k++) {
                i16 Aik = matrixA->data[i][k];
                i16 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_u8* matrixMultiplication_u8_c(Matrix_u8* matrixA, Matrix_u8* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_u8* c = makeMatrix_u8(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            u8 sum = 0;
            for (int k = 0 ; k < m; k++) {
                u8 Aik = matrixA->data[i][k];
                u8 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}
EMSCRIPTEN_KEEPALIVE
Matrix_i8* matrixMultiplication_i8_c(Matrix_i8* matrixA, Matrix_i8* matrixB) {
    int m = matrixA->width;
    int n = matrixA->height;
    int p = matrixB->width;
    Matrix_i8* c = makeMatrix_i8(n, p);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < p; j++) {
            i8 sum = 0;
            for (int k = 0 ; k < m; k++) {
                i8 Aik = matrixA->data[i][k];
                i8 Bkj = matrixB->data[k][j];
                sum += Aik * Bkj;
            }
            c->data[i][j] = sum;
        }
    }
    
    return c;
}

File: assemblyscript-algorithms/src/algorithms/matrix/matrix-multiplication.ts
export function matrixMultiplication_u64_asm(
  matrixA: StaticArray<StaticArray<u64>>,
  matrixB: StaticArray<StaticArray<u64>>
): StaticArray<StaticArray<u64>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<u64>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<u64>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: u64 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_i64_asm(
  matrixA: StaticArray<StaticArray<i64>>,
  matrixB: StaticArray<StaticArray<i64>>
): StaticArray<StaticArray<i64>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<i64>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<i64>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: i64 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_f64_asm(
  matrixA: StaticArray<StaticArray<f64>>,
  matrixB: StaticArray<StaticArray<f64>>
): StaticArray<StaticArray<f64>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<f64>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<f64>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: f64 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_f32_asm(
  matrixA: StaticArray<StaticArray<f32>>,
  matrixB: StaticArray<StaticArray<f32>>
): StaticArray<StaticArray<f32>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<f32>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<f32>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: f32 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_u32_asm(
  matrixA: StaticArray<StaticArray<u32>>,
  matrixB: StaticArray<StaticArray<u32>>
): StaticArray<StaticArray<u32>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<u32>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<u32>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: u32 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_i32_asm(
  matrixA: StaticArray<StaticArray<i32>>,
  matrixB: StaticArray<StaticArray<i32>>
): StaticArray<StaticArray<i32>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<i32>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<i32>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: i32 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_u16_asm(
  matrixA: StaticArray<StaticArray<u16>>,
  matrixB: StaticArray<StaticArray<u16>>
): StaticArray<StaticArray<u16>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<u16>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<u16>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: u16 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_i16_asm(
  matrixA: StaticArray<StaticArray<i16>>,
  matrixB: StaticArray<StaticArray<i16>>
): StaticArray<StaticArray<i16>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<i16>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<i16>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: i16 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_u8_asm(
  matrixA: StaticArray<StaticArray<u8>>,
  matrixB: StaticArray<StaticArray<u8>>
): StaticArray<StaticArray<u8>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<u8>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<u8>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: u8 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
export function matrixMultiplication_i8_asm(
  matrixA: StaticArray<StaticArray<i8>>,
  matrixB: StaticArray<StaticArray<i8>>
): StaticArray<StaticArray<i8>> {
  const m = matrixA[0].length;
  const n = matrixA.length;
  const p = matrixB[0].length;

  const c = new StaticArray<StaticArray<i8>>(n);
  for (let i = 0; i < n; i++) {
    c[i] = new StaticArray<i8>(p);
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < p; j++) {
      let sum: i8 = 0;
      for (let k = 0; k < m; k++) {
        const Aik = matrixA[i][k];
        const Bkj = matrixB[k][j];
        sum += Aik * Bkj;
      }
      c[i][j] = sum;
    }
  }

  return c;
}
